hkl/configuration.py

```diff
@@ -0,0 +1,454 @@
+        """
+        Restore configuration from a recognized format (dict, json, yaml).
+
+        Instead of guessing, recognize the kind of config data by its structure.
+
+        PARAMETERS
+
+        config *dict* or *str*:
+            structure (dict, json, or yaml) with diffractometer configuration
+        clear *bool*:
+            If ``True`` (default), remove any previous configuration of the
+            diffractometer and reset it to default values before restoring the
+            configuration.
+
+        Note: Can't name this method "import", it's a reserved Python word.
+        """
+        importer = None
+
+        if isinstance(config, dict):
+            importer = self.from_dict
+        elif isinstance(config, str):
+            if config.strip().startswith("{"):
+                importer = self.from_json
+            else:
+                importer = self.from_yaml
+        if importer is None:
+            raise TypeError("Unrecognized configuration structure.")
+
+        importer(config, clear=clear)
+
+    def reset_diffractometer(self):
+        """Clear all diffractometer settings."""
+        from .diffract import Diffractometer
+        from .util import Constraint
+
+        assert isinstance(self.diffractometer, Diffractometer)
+
+        self.diffractometer.wavelength = DEFAULT_WAVELENGTH
+        self.diffractometer.engine.mode = self.diffractometer.engine.modes[0]
+
+        # fmt: off
+        self.diffractometer._set_constraints(
+            {
+                k: Constraint(-180, 180, 0.0, True)
+                for k in self.real_axes_names
+            }
+        )
+        # fmt: on
+
+        for k in list(self.diffractometer.calc._samples):
+            self.diffractometer.calc._samples.pop(k)
+
+        # fmt: off
+        a0 = DEFAULT_WAVELENGTH  # coincidentally
+        self.diffractometer.calc.new_sample(
+            "main", lattice=(a0, a0, a0, 90, 90, 90)
+        )
+        # fmt: on
+
+    def validate_config_dict(self, config):
+        """
+        Verify that the config dictionary is correctly formed.
+
+        PARAMETERS
+
+        config *dict*:
+            structure (dict) with diffractometer configuration
+        """
+        from .diffract import Diffractometer
+        from .util import libhkl
+
+        diffractometer = self.diffractometer
+        assert isinstance(diffractometer, Diffractometer)
+        assert isinstance(config, dict)
+
+        for k, types in EXPECTED_CONFIGURATION_KEYS_TYPES.items():
+            assert k in config
+            assert isinstance(config[k], types)
+
+        calc = diffractometer.calc  # shortcut
+        assert config["canonical_axes"] == self.canonical_axes_names
+        assert config["engine"] == calc.engine.name
+        assert config["geometry"] == calc._geometry.name_get()
+        assert config["library"].split()[0] == libhkl.__name__
+        assert config["reciprocal_axes"] == self.reciprocal_axes_names
```

Comment:
```suggestion
        for k, types in EXPECTED_CONFIGURATION_KEYS_TYPES.items():
            assert k in config, f"Missing required parameter, {k}"
            assert isinstance(config[k], types), f"Bad value type for parameter, {k}"

        calc = diffractometer.calc  # shortcut
        assert config["canonical_axes"] == self.canonical_axes_names
        assert config["engine"] == calc.engine.name
        assert config["geometry"] == calc._geometry.name_get()
        assert config["library"].split()[0] == libhkl.__name__
        assert config["reciprocal_axes"] == self.reciprocal_axes_names
```

Add helpful error messages